sympy/solvers/solveset.py

```diff
@@ -1121,3 +1126,597 @@ def linsolve(system, *symbols):
+        old_result, solveset_complex)
+
+    # overall result
+    result = new_result_real + new_result_complex
+    result_finiteset = []
+    for res in result:
+        if not res:
+            # means {None : None}
+            continue
+        # If length < len(all_symbols) means infinite soln.
+        # Some or all the soln is dependent on 1 symbol.
+        # eg. {x: y+2} then final soln {x: y+2, y: y}
+        if len(res) < len(all_symbols):
+            solved_symbols = res.keys()
+            unsolved = list(filter(
+                lambda x: x not in solved_symbols, all_symbols))
+            rcopy = res.copy()
+            for unsolved_sym in unsolved:
+                rcopy[unsolved_sym] = unsolved_sym
+            res = rcopy
+        result_finiteset.append(res)
+
+    if intersections:
+        # If solveset have returned some intersection for any symbol
+        # intersection is Interval or Set
+        result = []
+        for res in result_finiteset:
+            res_copy = res
+            for key_res, value_res in res.items():
+                for key_intersec, value_intersec in intersections.items():
+                    if key_intersec == key_res:
+                        res_copy[key_res] = \
+                            Intersection(FiniteSet(value_res), value_intersec)
+            result.append(res_copy)
+        result_finiteset = result
+
+    if complements:
+        # If solveset have returned some complements for any symbol
+        # complements are Set.
+        result = []
+        for res in result_finiteset:
+            res_copy = res
+            for key_res, value_res in res.items():
+                for key_complement, value_complement in complements.items():
+                    if key_complement == key_res:
+                        res_copy[key_res] = \
+                            FiniteSet(value_res) - value_complement
+            result.append(res_copy)
+        result_finiteset = result
+
+    # convert to ordered tuple
+    result = S.EmptySet
+    for r in result_finiteset:
+        temp = [r[symb] for symb in all_symbols]
+        result += FiniteSet(tuple(temp))
+    return result
+
+
+def nonlinsolve(system, symbols):
+    r"""
+    Solve system of N non linear equations with M variables, which means both
+    under and overdetermined systems are supported. Positive dimensional
+    system is also supported (A system with infinitely many solutions is said
+    to be positive-dimensional). In Positive dimensional system solution will
+    be dependent on at least one symbol. Returns both real solution
+    and complex solution. The possible number of solutions is zero,
+    one or infinite.
+
+    Parameters
+    ==========
+
+    system : list of equations
+        The target system of equations
+    symbols : list of Symbols
+        symbols should be given as a sequence eg. list
+
+    Returns
+    =======
+
+    A FiniteSet of ordered tuple of values of `symbols` for which the `system`
```

Comment:
> I followed these word

@Shekharrajak: Just because something is written somewhere doesn't mean it is correct, please use your own judgment to justify the code you write.

Though saying `ordered tuple` is redundant, it might add clarity for users who are not very well versed with python. So I'll say keep it.